File: scripts/kospi_top10_analysis.py
#!/usr/bin/env python3
"""KOSPI TOP 10 — 일봉 지표 전일 대비 등락비율(A안) CSV 생성."""

from __future__ import annotations

import csv
import os
import sys
from datetime import date, datetime, timedelta, timezone
from typing import Any

import yfinance as yf

ROOT = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))

KOSPI_TOP_10: list[tuple[str, str]] = [
    ("005930.KS", "삼성전자"),
    ("000660.KS", "SK하이닉스"),
    ("373220.KS", "LG에너지솔루션"),
    ("207940.KS", "삼성바이오로직스"),
    ("005380.KS", "현대차"),
    ("329180.KS", "HD현대중공업"),
    ("000270.KS", "기아"),
    ("105560.KS", "KB금융"),
    ("035420.KS", "NAVER"),
    ("055550.KS", "신한지주"),
]

OUTPUT_MONTHS = 2
FETCH_PERIOD = "5mo"
CSV_HEADERS = [
    "종목코드",
    "종목명",
    "일자",
    "종가",
    "SMA5_등락비율",
    "SMA20_등락비율",
    "SMA60_등락비율",
    "BB상_등락비율",
    "BB중_등락비율",
    "BB하_등락비율",
    "RSI_등락비율",
    "MACD_등락비율",
]


def pct_change(curr: float | None, prev: float | None) -> str:
    if curr is None or prev is None or prev == 0:
        return ""
    return f"{((curr / prev) - 1) * 100:.4f}"


def sma_series(closes: list[float], period: int) -> list[float | None]:
    out: list[float | None] = [None] * len(closes)
    for i in range(period - 1, len(closes)):
        window = closes[i - period + 1 : i + 1]
        out[i] = sum(window) / period
    return out


def bollinger_series(closes: list[float], period: int = 20, mult: float = 2.0) -> tuple[list[float | None], list[float | None], list[float | None]]:
    upper: list[float | None] = [None] * len(closes)
    middle: list[float | None] = [None] * len(closes)
    lower: list[float | None] = [None] * len(closes)
    for i in range(period - 1, len(closes)):
        window = closes[i - period + 1 : i + 1]
        mean = sum(window) / period
        variance = sum((x - mean) ** 2 for x in window) / period
        std = variance**0.5
        middle[i] = mean
        upper[i] = mean + mult * std
        lower[i] = mean - mult * std
    return upper, middle, lower


def rsi_series(closes: list[float], period: int = 14) -> list[float | None]:
    out: list[float | None] = [None] * len(closes)
    if len(closes) <= period:
        return out

    gains = 0.0
    losses = 0.0
    for i in range(1, period + 1):
        diff = closes[i] - closes[i - 1]
        if diff >= 0:
            gains += diff
        else:
            losses -= diff
    avg_gain = gains / period
    avg_loss = losses / period
    rs = 100.0 if avg_loss == 0 else avg_gain / avg_loss
    out[period] = 100 - 100 / (1 + rs)

    for i in range(period + 1, len(closes)):
        diff = closes[i] - closes[i - 1]
        gain = diff if diff > 0 else 0.0
        loss = -diff if diff < 0 else 0.0
        avg_gain = (avg_gain * (period - 1) + gain) / period
        avg_loss = (avg_loss * (period - 1) + loss) / period
        rs_val = 100.0 if avg_loss == 0 else avg_gain / avg_loss
        out[i] = 100 - 100 / (1 + rs_val)
    return out


def ema_array(data: list[float], period: int) -> list[float | None]:
    arr: list[float | None] = [None] * len(data)
    k = 2 / (period + 1)
    prev: float | None = None
    for i in range(len(data)):
        if i < period - 1:
            continue
        if prev is None:
            prev = sum(data[i - period + 1 : i + 1]) / period
        else:
            prev = data[i] * k + prev * (1 - k)
        arr[i] = prev
    return arr


def macd_series(closes: list[float], fast: int = 12, slow: int = 26, signal_period: int = 9) -> list[float | None]:
    out: list[float | None] = [None] * len(closes)
    fast_arr = ema_array(closes, fast)
    slow_arr = ema_array(closes, slow)
    macd_line: list[float] = []
    macd_idx: list[int] = []
    for i in range(len(closes)):
        if fast_arr[i] is not None and slow_arr[i] is not None:
            macd_line.append(fast_arr[i] - slow_arr[i])  # type: ignore[operator]
            macd_idx.append(i)
    if not macd_line:
        return out

    k = 2 / (signal_period + 1)
    sig_prev: float | None = None
    signal_values: list[tuple[int, float]] = []
    for j, val in enumerate(macd_line):
        if j < signal_period - 1:
            continue
        if sig_prev is None:
            sig_prev = sum(macd_line[j - signal_period + 1 : j + 1]) / signal_period
        else:
            sig_prev = val * k + sig_prev * (1 - k)
        signal_values.append((macd_idx[j], val))

    for idx, val in signal_values:
        out[idx] = val
    return out


def fetch_candles(ticker: str) -> list[dict[str, Any]]:
    hist = yf.Ticker(ticker).history(period=FETCH_PERIOD, interval="1d", auto_adjust=False)
    if hist is None or hist.empty:
        return []

    candles: list[dict[str, Any]] = []
    for idx, row in hist.iterrows():
        close = row.get("Close")
        if close is None or (hasattr(close, "__float__") and float(close) != float(close)):
            continue
        close_f = float(close)
        if hasattr(idx, "date"):
            day = idx.date()
        elif hasattr(idx, "to_pydatetime"):
            day = idx.to_pydatetime().date()
        else:
            day = datetime.fromisoformat(str(idx)[:10]).date()
        candles.append({"date": day, "close": round(close_f, 4)})
    return candles


def build_rows(ticker: str, name: str, cutoff: date) -> list[dict[str, str]]:
    candles = fetch_candles(ticker)
    if not candles:
        return []

    dates = [c["date"] for c in candles]
    closes = [c["close"] for c in candles]

    sma5 = sma_series(closes, 5)
    sma20 = sma_series(closes, 20)
    sma60 = sma_series(closes, 60)
    bb_u, bb_m, bb_l = bollinger_series(closes, 20, 2.0)
    rsi = rsi_series(closes, 14)
    macd = macd_series(closes, 12, 26, 9)

    rows: list[dict[str, str]] = []
    for i in range(len(candles)):
        if dates[i] < cutoff:
            continue
        prev = i - 1
        row = {
            "종목코드": ticker,
            "종목명": name,
            "일자": dates[i].isoformat(),
            "종가": f"{closes[i]:.4f}",
            "SMA5_등락비율": pct_change(sma5[i], sma5[prev] if prev >= 0 else None),
            "SMA20_등락비율": pct_change(sma20[i], sma20[prev] if prev >= 0 else None),
            "SMA60_등락비율": pct_change(sma60[i], sma60[prev] if prev >= 0 else None),
            "BB상_등락비율": pct_change(bb_u[i], bb_u[prev] if prev >= 0 else None),
            "BB중_등락비율": pct_change(bb_m[i], bb_m[prev] if prev >= 0 else None),
            "BB하_등락비율": pct_change(bb_l[i], bb_l[prev] if prev >= 0 else None),
            "RSI_등락비율": pct_change(rsi[i], rsi[prev] if prev >= 0 else None),
            "MACD_등락비율": pct_change(macd[i], macd[prev] if prev >= 0 else None),
        }
        rows.append(row)
    return rows


def main() -> None:
    today = datetime.now(timezone.utc).date()
    cutoff = today - timedelta(days=OUTPUT_MONTHS * 31)

    out_dir = os.path.join(ROOT, "data")
    os.makedirs(out_dir, exist_ok=True)
    out_path = os.path.join(out_dir, "kospi-top10-analysis.csv")

    all_rows: list[dict[str, str]] = []
    for ticker, name in KOSPI_TOP_10:
        print(f"Analyzing {ticker} {name}...")
        rows = build_rows(ticker, name, cutoff)
        print(f"  -> {len(rows)} rows")
        all_rows.extend(rows)

    all_rows.sort(key=lambda r: (r["종목코드"], r["일자"]))

    with open(out_path, "w", encoding="utf-8-sig", newline="") as handle:
        writer = csv.DictWriter(handle, fieldnames=CSV_HEADERS)
        writer.writeheader()
        writer.writerows(all_rows)

    print(f"Wrote {out_path} ({len(all_rows)} rows, since {cutoff.isoformat()})")


if __name__ == "__main__":
    main()

File: scripts/kospi_top10_analysis.mjs
#!/usr/bin/env node
/** KOSPI TOP 10 — 최근 2개월 지표 전일 대비 등락비율(A안) CSV */

import fs from "fs";
import path from "path";
import { fileURLToPath } from "url";

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const ROOT = path.resolve(__dirname, "..");
const API_BASE = (process.env.STOCK_API_URL || "https://first-stock-api.onrender.com").replace(/\/$/, "");
const OUTPUT_MONTHS = 2;
const CUTOFF = new Date();
CUTOFF.setUTCDate(CUTOFF.getUTCDate() - OUTPUT_MONTHS * 31);
CUTOFF.setUTCHours(0, 0, 0, 0);

const KOSPI_TOP_10 = [
  ["005930.KS", "삼성전자"],
  ["000660.KS", "SK하이닉스"],
  ["373220.KS", "LG에너지솔루션"],
  ["207940.KS", "삼성바이오로직스"],
  ["005380.KS", "현대차"],
  ["329180.KS", "HD현대중공업"],
  ["000270.KS", "기아"],
  ["105560.KS", "KB금융"],
  ["035420.KS", "NAVER"],
  ["055550.KS", "신한지주"]
];

const HEADERS = [
  "종목코드",
  "종목명",
  "일자",
  "종가",
  "SMA5_등락비율",
  "SMA20_등락비율",
  "SMA60_등락비율",
  "BB상_등락비율",
  "BB중_등락비율",
  "BB하_등락비율",
  "RSI_등락비율",
  "MACD_등락비율"
];

function pctChange(curr, prev) {
  if (curr == null || prev == null || prev === 0) return "";
  return (((curr / prev) - 1) * 100).toFixed(4);
}

function sma(closes, period) {
  const out = new Array(closes.length).fill(null);
  for (let i = period - 1; i < closes.length; i += 1) {
    let sum = 0;
    for (let j = i - period + 1; j <= i; j += 1) sum += closes[j];
    out[i] = sum / period;
  }
  return out;
}

function bollinger(closes, period = 20, mult = 2) {
  const upper = new Array(closes.length).fill(null);
  const middle = new Array(closes.length).fill(null);
  const lower = new Array(closes.length).fill(null);
  for (let i = period - 1; i < closes.length; i += 1) {
    let sum = 0;
    for (let j = i - period + 1; j <= i; j += 1) sum += closes[j];
    const mean = sum / period;
    let variance = 0;
    for (let j = i - period + 1; j <= i; j += 1) variance += (closes[j] - mean) ** 2;
    const std = Math.sqrt(variance / period);
    middle[i] = mean;
    upper[i] = mean + mult * std;
    lower[i] = mean - mult * std;
  }
  return { upper, middle, lower };
}

function rsi(closes, period = 14) {
  const out = new Array(closes.length).fill(null);
  if (closes.length <= period) return out;
  let gains = 0;
  let losses = 0;
  for (let i = 1; i <= period; i += 1) {
    const diff = closes[i] - closes[i - 1];
    if (diff >= 0) gains += diff;
    else losses -= diff;
  }
  let avgGain = gains / period;
  let avgLoss = losses / period;
  let rs = avgLoss === 0 ? 100 : avgGain / avgLoss;
  out[period] = 100 - 100 / (1 + rs);
  for (let i = period + 1; i < closes.length; i += 1) {
    const diff = closes[i] - closes[i - 1];
    const gain = diff > 0 ? diff : 0;
    const loss = diff < 0 ? -diff : 0;
    avgGain = (avgGain * (period - 1) + gain) / period;
    avgLoss = (avgLoss * (period - 1) + loss) / period;
    rs = avgLoss === 0 ? 100 : avgGain / avgLoss;
    out[i] = 100 - 100 / (1 + rs);
  }
  return out;
}

function emaArray(data, period) {
  const arr = new Array(data.length).fill(null);
  const k = 2 / (period + 1);
  let prev = null;
  for (let i = 0; i < data.length; i += 1) {
    if (i < period - 1) continue;
    if (prev == null) {
      let sum = 0;
      for (let j = i - period + 1; j <= i; j += 1) sum += data[j];
      prev = sum / period;
    } else {
      prev = data[i] * k + prev * (1 - k);
    }
    arr[i] = prev;
  }
  return arr;
}

function macdLine(closes) {
  const out = new Array(closes.length).fill(null);
  const fast = emaArray(closes, 12);
  const slow = emaArray(closes, 26);
  const macdIdx = [];
  const macdVals = [];
  for (let i = 0; i < closes.length; i += 1) {
    if (fast[i] != null && slow[i] != null) {
      macdIdx.push(i);
      macdVals.push(fast[i] - slow[i]);
    }
  }
  const signalPeriod = 9;
  const k = 2 / (signalPeriod + 1);
  let sigPrev = null;
  for (let j = 0; j < macdVals.length; j += 1) {
    if (j < signalPeriod - 1) continue;
    const val = macdVals[j];
    if (sigPrev == null) {
      let sum = 0;
      for (let x = j - signalPeriod + 1; x <= j; x += 1) sum += macdVals[x];
      sigPrev = sum / signalPeriod;
    } else {
      sigPrev = val * k + sigPrev * (1 - k);
    }
    out[macdIdx[j]] = val;
  }
  return out;
}

async function fetchCandles(ticker) {
  const url = `${API_BASE}/api/chart?ticker=${encodeURIComponent(ticker)}&period=3mo&interval=1d`;
  const res = await fetch(url, { signal: AbortSignal.timeout(180000) });
  if (!res.ok) throw new Error(`${ticker} HTTP ${res.status}`);
  const data = await res.json();
  return data.candles || [];
}

function csvEscape(value) {
  const s = String(value ?? "");
  if (/[",\n]/.test(s)) return `"${s.replace(/"/g, '""')}"`;
  return s;
}

function buildRows(ticker, name, candles) {
  const dates = candles.map((c) => new Date(`${c.time}T00:00:00Z`));
  const closes = candles.map((c) => c.close);
  const sma5 = sma(closes, 5);
  const sma20 = sma(closes, 20);
  const sma60 = sma(closes, 60);
  const bb = bollinger(closes, 20, 2);
  const rsiVals = rsi(closes, 14);
  const macdVals = macdLine(closes);
  const rows = [];

  for (let i = 0; i < dates.length; i += 1) {
    if (dates[i] < CUTOFF) continue;
    const prev = i - 1;
    rows.push({
      종목코드: ticker,
      종목명: name,
      일자: candles[i].time,
      종가: closes[i].toFixed(4),
      SMA5_등락비율: pctChange(sma5[i], prev >= 0 ? sma5[prev] : null),
      SMA20_등락비율: pctChange(sma20[i], prev >= 0 ? sma20[prev] : null),
      SMA60_등락비율: pctChange(sma60[i], prev >= 0 ? sma60[prev] : null),
      BB상_등락비율: pctChange(bb.upper[i], prev >= 0 ? bb.upper[prev] : null),
      BB중_등락비율: pctChange(bb.middle[i], prev >= 0 ? bb.middle[prev] : null),
      BB하_등락비율: pctChange(bb.lower[i], prev >= 0 ? bb.lower[prev] : null),
      RSI_등락비율: pctChange(rsiVals[i], prev >= 0 ? rsiVals[prev] : null),
      MACD_등락비율: pctChange(macdVals[i], prev >= 0 ? macdVals[prev] : null)
    });
  }
  return rows;
}

async function main() {
  const allRows = [];
  for (const [ticker, name] of KOSPI_TOP_10) {
    process.stdout.write(`Fetching ${ticker} ${name}...\n`);
    const candles = await fetchCandles(ticker);
    const rows = buildRows(ticker, name, candles);
    process.stdout.write(`  -> ${rows.length} rows\n`);
    allRows.push(...rows);
  }

  allRows.sort((a, b) => (a.종목코드 === b.종목코드 ? a.일자.localeCompare(b.일자) : a.종목코드.localeCompare(b.종목코드)));

  const lines = ["\uFEFF" + HEADERS.map(csvEscape).join(",")];
  for (const row of allRows) {
    lines.push(HEADERS.map((h) => csvEscape(row[h])).join(","));
  }

  const outDir = path.join(ROOT, "data");
  fs.mkdirSync(outDir, { recursive: true });
  const outPath = path.join(outDir, "kospi-top10-analysis.csv");
  fs.writeFileSync(outPath, lines.join("\n"), "utf8");
  process.stdout.write(`Wrote ${outPath} (${allRows.length} rows, since ${CUTOFF.toISOString().slice(0, 10)})\n`);
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});
